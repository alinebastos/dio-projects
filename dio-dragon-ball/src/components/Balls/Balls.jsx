import React, { useState } from 'react';

import {
  Card as BCard,
  CardImg,
  CardBody,
  CardTitle,
  CardSubtitle,
  Badge,
  Col,
  Row,
  Form,
  FormGroup,
  Label,
  Button,
  Input,
  Modal,
  ModalFooter,
  ModalBody,
} from 'reactstrap';
import { Text, Flex } from 'rebass';

import styled from 'styled-components';

const Card = styled(BCard)`
  height: 280px;
  margin: 5px 0px;
`;

const Balls = ({ balls, profile }) => {
  const [list, setList] = useState(balls);
  const [modal, setModal] = useState(false);
  const [currentBall, setBall] = useState(null);

  const toggle = () => setModal(!modal);

  const validateBall = (ball) => {
    setBall(ball)
    setModal(true)
  }

  const filterByMe = () => {
    return balls.filter((ball) => ball.owner === profile.id);
  };

  const filterNotMe = () => {
    return balls.filter((ball) => ball.owner !== profile.id);
  };

  const filter = (value) => {
    const cases = {
      me: () => setList(filterByMe()),
      all: () => setList(balls),
      notme: () => setList(filterNotMe()),
    };
    return cases[value]();
  };

  const updateList = (id) => {
    const newList = balls.map((ball) => {
      if(ball.id === id) return {
        ...ball,
        owner: profile.id
      }
      return ball
    })
    const newProfile = profile.balls.push(id)

    setList(newList)
    setModal(false)
  }

  return (
    <>
      <Flex my='10px;' justifyContent='space-between'>
        <Text fontSize='50px'>Esferas</Text>
        <FormGroup>
          <Label for='filter'>Filtrar</Label>
          <Input
            type='select'
            name='select'
            id='filter'
            data-testid="filter"
            onChange={({ target: { value } }) => filter(value)}
          >
            <option value='all'>Todas as esferas</option>
            <option value='me'>Minhas esferas</option>
            <option value='notme'>Não tenho</option>
          </Input>
        </FormGroup>
      </Flex>
      <Row>
        {list.length > 0 ? (
          list.map((ball, i) => (
            <Col sm='3' key={ball.id}>
              <Card>
                <CardImg top width='100%' src={ball.image} alt={ball.name} />
                <CardBody>
                  <CardTitle>{ball.name}</CardTitle>
                  {ball.owner !== profile.id ? (
                    <>
                      <Badge color='danger'>Não encontrada</Badge>
                      <Button size='sm' color='warning' onClick={() => validateBall(ball)}>
                        encontrei
                      </Button>
                    </>
                  ) : (
                    <Badge color='success'>Encontrada</Badge>
                  )}
                </CardBody>
              </Card>
            </Col>
          ))
        ) : (
          <p>Houve algum problema na busca das esferas</p>
        )}
      </Row>
      {/* Validate ball */}
      <Modal isOpen={modal} toggle={toggle}>
        <ModalBody>
          <FormGroup>
            <Label for='code'>Insira o código da esfera de {currentBall?.name}:</Label>
            <Input
              type='number'
              name='ballcode'
              id='code'
              placeholder='Ex: 23412'
            />
          </FormGroup>
        </ModalBody>
        <ModalFooter>
        <Button color='success' onClick={() => updateList(currentBall?.id)}>
            Validar
          </Button>
          <Button color='secondary' onClick={toggle}>
            Voltar
          </Button>
        </ModalFooter>
      </Modal>
    </>
  );
};

export default Balls;
