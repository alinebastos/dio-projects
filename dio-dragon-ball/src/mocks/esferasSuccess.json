
{
  "balls": [
    {
      "id":1,
      "name": "1 estrela",
      "owner": 1,
      "image": "https://i.pinimg.com/564x/a5/f9/a2/a5f9a2eb5c0bfb1f66988696e1f31334.jpg"
    },
    {
      "id":2,
      "name": "2 estrelas",
      "owner": 1,
      "image": "https://i.pinimg.com/564x/0b/72/75/0b72755a5ea5e1999d6fbfd3aa544355.jpg"
    },
    {
      "id":3,
      "name": "3 estrelas",
      "owner": 1,
      "image": "https://i.pinimg.com/564x/32/eb/b8/32ebb8536d0fa4108eef460694586d39.jpg"
    },
    {
      "id":4,
      "name": "4 estrelas",
      "owner": 1,
      "image": "https://i.pinimg.com/564x/7a/7f/76/7a7f76ea01702078582a1ae8b7344fb3.jpg"
    },
    {
      "id":5,
      "name": "5 estrelas",
      "owner": 1,
      "image": "https://i.pinimg.com/564x/13/19/62/1319628426fa9d3783705daf7db525f8.jpg"
    },
    {
      "id":6,
      "name": "6 estrelas",
      "owner": 1,
      "image": "https://i.pinimg.com/originals/6c/72/47/6c7247dfb67e18add93d682dc9fdabcc.png"
    },
    {
      "id":7,
      "name": "7 estrelas",
      "owner": 1,
      "image":"https://i.pinimg.com/564x/ec/8e/c5/ec8ec55eb1b4d4dfa3f7b2cbffe5bc6c.jpg"
    }
  ]
}
