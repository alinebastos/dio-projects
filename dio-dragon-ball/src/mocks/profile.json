{
  "profile": {
    "id": 1,
    "name": "La√≠s Lima",
    "image": "https://pbs.twimg.com/profile_images/1252965332785991680/8Ry7z6Pt_400x400.jpg",
    "balls": [1, 5, 6, 7]
  }
}
