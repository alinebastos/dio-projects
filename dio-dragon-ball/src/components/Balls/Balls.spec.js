import React from 'react'
import { render, fireEvent, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import Balls from './Balls'

//mocks
import { profile } from '../../mocks/profile.json';
import profileSuccess from '../../mocks/profileSuccess.json';
import esferas from '../../mocks/esferas.json';
import esferasSuccess from '../../mocks/esferasSuccess.json';

it('Should render all dragon balls', () => {
  const { container } = render(<Balls balls={esferas.balls} profile={profile} />)
  expect(container).toMatchSnapshot()
})

it('Should show not found balls label and button', () => {
  const { getAllByText } = render(<Balls balls={esferas.balls} profile={profile} />)

  const notFound = !!(getAllByText('Não encontrada').length > 0)
  const validateButton = !!(getAllByText('encontrei').length > 0)

  expect(notFound).toBeTruthy()
  expect(validateButton).toBeTruthy()
})

it('Should open modal to validate', () => {
  const { debug, getAllByText, getByText, container } = render(<Balls balls={esferas.balls} profile={profile} />)
  const validateButton = getAllByText('encontrei')

  fireEvent.click(validateButton[0])

  expect(getByText('Validar')).toBeTruthy()
  expect(container).toMatchSnapshot()

})

it('Should filter', async () => {
  const { debug, getByTestId, getByText, getAllByText, container } = render(<Balls balls={esferas.balls} profile={profile} />)
  const filter = getByTestId('filter')
  fireEvent.click(filter)

  const myDragonBalls = getByText('Não tenho').closest('option')
  fireEvent.click(myDragonBalls)

  await waitFor(() => {
    expect(getAllByText('Não encontrada').length).toBe(3)
    // console.log(getAllByText('Encontrada').length)
    // expect(getAllByText('Encontrada')).toBe(3).not.toBeInTheDocument()
  })
})
