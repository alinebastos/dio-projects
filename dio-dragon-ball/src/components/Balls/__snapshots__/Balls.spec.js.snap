// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should open modal to validate 1`] = `
<div>
  <div
    class="css-mv96ze"
  >
    <div
      class="css-1thq8i2"
    >
      Esferas
    </div>
    <div
      class="form-group"
    >
      <label
        class=""
        for="filter"
      >
        Filtrar
      </label>
      <select
        class="form-control"
        data-testid="filter"
        id="filter"
        name="select"
      >
        <option
          value="all"
        >
          Todas as esferas
        </option>
        <option
          value="me"
        >
          Minhas esferas
        </option>
        <option
          value="notme"
        >
          Não tenho
        </option>
      </select>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="col-sm-3"
    >
      <div
        class="sc-AxjAm iGDTZF card"
      >
        <img
          alt="1 estrela"
          class="card-img-top"
          src="https://i.pinimg.com/564x/a5/f9/a2/a5f9a2eb5c0bfb1f66988696e1f31334.jpg"
          width="100%"
        />
        <div
          class="card-body"
        >
          <div
            class="card-title"
          >
            1 estrela
          </div>
          <span
            class="badge badge-success"
          >
            Encontrada
          </span>
        </div>
      </div>
    </div>
    <div
      class="col-sm-3"
    >
      <div
        class="sc-AxjAm iGDTZF card"
      >
        <img
          alt="2 estrelas"
          class="card-img-top"
          src="https://i.pinimg.com/564x/0b/72/75/0b72755a5ea5e1999d6fbfd3aa544355.jpg"
          width="100%"
        />
        <div
          class="card-body"
        >
          <div
            class="card-title"
          >
            2 estrelas
          </div>
          <span
            class="badge badge-danger"
          >
            Não encontrada
          </span>
          <button
            class="btn btn-warning btn-sm"
            type="button"
          >
            encontrei
          </button>
        </div>
      </div>
    </div>
    <div
      class="col-sm-3"
    >
      <div
        class="sc-AxjAm iGDTZF card"
      >
        <img
          alt="3 estrelas"
          class="card-img-top"
          src="https://i.pinimg.com/564x/32/eb/b8/32ebb8536d0fa4108eef460694586d39.jpg"
          width="100%"
        />
        <div
          class="card-body"
        >
          <div
            class="card-title"
          >
            3 estrelas
          </div>
          <span
            class="badge badge-danger"
          >
            Não encontrada
          </span>
          <button
            class="btn btn-warning btn-sm"
            type="button"
          >
            encontrei
          </button>
        </div>
      </div>
    </div>
    <div
      class="col-sm-3"
    >
      <div
        class="sc-AxjAm iGDTZF card"
      >
        <img
          alt="4 estrelas"
          class="card-img-top"
          src="https://i.pinimg.com/564x/7a/7f/76/7a7f76ea01702078582a1ae8b7344fb3.jpg"
          width="100%"
        />
        <div
          class="card-body"
        >
          <div
            class="card-title"
          >
            4 estrelas
          </div>
          <span
            class="badge badge-danger"
          >
            Não encontrada
          </span>
          <button
            class="btn btn-warning btn-sm"
            type="button"
          >
            encontrei
          </button>
        </div>
      </div>
    </div>
    <div
      class="col-sm-3"
    >
      <div
        class="sc-AxjAm iGDTZF card"
      >
        <img
          alt="5 estrelas"
          class="card-img-top"
          src="https://i.pinimg.com/564x/13/19/62/1319628426fa9d3783705daf7db525f8.jpg"
          width="100%"
        />
        <div
          class="card-body"
        >
          <div
            class="card-title"
          >
            5 estrelas
          </div>
          <span
            class="badge badge-success"
          >
            Encontrada
          </span>
        </div>
      </div>
    </div>
    <div
      class="col-sm-3"
    >
      <div
        class="sc-AxjAm iGDTZF card"
      >
        <img
          alt="6 estrelas"
          class="card-img-top"
          src="https://i.pinimg.com/originals/6c/72/47/6c7247dfb67e18add93d682dc9fdabcc.png"
          width="100%"
        />
        <div
          class="card-body"
        >
          <div
            class="card-title"
          >
            6 estrelas
          </div>
          <span
            class="badge badge-success"
          >
            Encontrada
          </span>
        </div>
      </div>
    </div>
    <div
      class="col-sm-3"
    >
      <div
        class="sc-AxjAm iGDTZF card"
      >
        <img
          alt="7 estrelas"
          class="card-img-top"
          src="https://i.pinimg.com/564x/ec/8e/c5/ec8ec55eb1b4d4dfa3f7b2cbffe5bc6c.jpg"
          width="100%"
        />
        <div
          class="card-body"
        >
          <div
            class="card-title"
          >
            7 estrelas
          </div>
          <span
            class="badge badge-success"
          >
            Encontrada
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should render all dragon balls 1`] = `
<div>
  <div
    class="css-mv96ze"
  >
    <div
      class="css-1thq8i2"
    >
      Esferas
    </div>
    <div
      class="form-group"
    >
      <label
        class=""
        for="filter"
      >
        Filtrar
      </label>
      <select
        class="form-control"
        data-testid="filter"
        id="filter"
        name="select"
      >
        <option
          value="all"
        >
          Todas as esferas
        </option>
        <option
          value="me"
        >
          Minhas esferas
        </option>
        <option
          value="notme"
        >
          Não tenho
        </option>
      </select>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="col-sm-3"
    >
      <div
        class="sc-AxjAm iGDTZF card"
      >
        <img
          alt="1 estrela"
          class="card-img-top"
          src="https://i.pinimg.com/564x/a5/f9/a2/a5f9a2eb5c0bfb1f66988696e1f31334.jpg"
          width="100%"
        />
        <div
          class="card-body"
        >
          <div
            class="card-title"
          >
            1 estrela
          </div>
          <span
            class="badge badge-success"
          >
            Encontrada
          </span>
        </div>
      </div>
    </div>
    <div
      class="col-sm-3"
    >
      <div
        class="sc-AxjAm iGDTZF card"
      >
        <img
          alt="2 estrelas"
          class="card-img-top"
          src="https://i.pinimg.com/564x/0b/72/75/0b72755a5ea5e1999d6fbfd3aa544355.jpg"
          width="100%"
        />
        <div
          class="card-body"
        >
          <div
            class="card-title"
          >
            2 estrelas
          </div>
          <span
            class="badge badge-danger"
          >
            Não encontrada
          </span>
          <button
            class="btn btn-warning btn-sm"
            type="button"
          >
            encontrei
          </button>
        </div>
      </div>
    </div>
    <div
      class="col-sm-3"
    >
      <div
        class="sc-AxjAm iGDTZF card"
      >
        <img
          alt="3 estrelas"
          class="card-img-top"
          src="https://i.pinimg.com/564x/32/eb/b8/32ebb8536d0fa4108eef460694586d39.jpg"
          width="100%"
        />
        <div
          class="card-body"
        >
          <div
            class="card-title"
          >
            3 estrelas
          </div>
          <span
            class="badge badge-danger"
          >
            Não encontrada
          </span>
          <button
            class="btn btn-warning btn-sm"
            type="button"
          >
            encontrei
          </button>
        </div>
      </div>
    </div>
    <div
      class="col-sm-3"
    >
      <div
        class="sc-AxjAm iGDTZF card"
      >
        <img
          alt="4 estrelas"
          class="card-img-top"
          src="https://i.pinimg.com/564x/7a/7f/76/7a7f76ea01702078582a1ae8b7344fb3.jpg"
          width="100%"
        />
        <div
          class="card-body"
        >
          <div
            class="card-title"
          >
            4 estrelas
          </div>
          <span
            class="badge badge-danger"
          >
            Não encontrada
          </span>
          <button
            class="btn btn-warning btn-sm"
            type="button"
          >
            encontrei
          </button>
        </div>
      </div>
    </div>
    <div
      class="col-sm-3"
    >
      <div
        class="sc-AxjAm iGDTZF card"
      >
        <img
          alt="5 estrelas"
          class="card-img-top"
          src="https://i.pinimg.com/564x/13/19/62/1319628426fa9d3783705daf7db525f8.jpg"
          width="100%"
        />
        <div
          class="card-body"
        >
          <div
            class="card-title"
          >
            5 estrelas
          </div>
          <span
            class="badge badge-success"
          >
            Encontrada
          </span>
        </div>
      </div>
    </div>
    <div
      class="col-sm-3"
    >
      <div
        class="sc-AxjAm iGDTZF card"
      >
        <img
          alt="6 estrelas"
          class="card-img-top"
          src="https://i.pinimg.com/originals/6c/72/47/6c7247dfb67e18add93d682dc9fdabcc.png"
          width="100%"
        />
        <div
          class="card-body"
        >
          <div
            class="card-title"
          >
            6 estrelas
          </div>
          <span
            class="badge badge-success"
          >
            Encontrada
          </span>
        </div>
      </div>
    </div>
    <div
      class="col-sm-3"
    >
      <div
        class="sc-AxjAm iGDTZF card"
      >
        <img
          alt="7 estrelas"
          class="card-img-top"
          src="https://i.pinimg.com/564x/ec/8e/c5/ec8ec55eb1b4d4dfa3f7b2cbffe5bc6c.jpg"
          width="100%"
        />
        <div
          class="card-body"
        >
          <div
            class="card-title"
          >
            7 estrelas
          </div>
          <span
            class="badge badge-success"
          >
            Encontrada
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
